atl-checker/src/algorithms/certain_zero/search_strategy/linear_optimize.rs

```diff
@@ -0,0 +1,505 @@
+
+                return RangedPhi::Or(
+                    Box::from(pre_symbol_range_map),
+                    Box::from(until_symbol_range_map),
+                );
+            }
+            Phi::DespiteEventually { formula, .. } => {
+                return self.map_phi_to_ranges(formula);
+            }
+            Phi::EnforceEventually { formula, .. } => {
+                return self.map_phi_to_ranges(formula);
+            }
+            Phi::DespiteInvariant { formula, .. } => {
+                return self.map_phi_to_ranges(formula);
+            }
+            Phi::EnforceInvariant { formula, .. } => {
+                return self.map_phi_to_ranges(formula);
+            }
+        }
+    }
+
+    /// Goes through the RangedPhi and finds how close we are to acceptance border in this state.
+    fn visit_ranged_phi(&self, ranged_phi: &RangedPhi, state: &State) -> Option<i32> {
+        match ranged_phi {
+            // If we need to satisfy either of the formulas, just return the lowest distance found between the two
+            RangedPhi::Or(lhs, rhs) => {
+                if let Some(lhs_distance) = self.visit_ranged_phi(lhs, state) {
+                    return if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
+                        Some(Ord::min(lhs_distance, rhs_distance))
+                    } else {
+                        Some(lhs_distance)
+                    };
+                } else {
+                    if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
+                        return Some(rhs_distance);
+                    }
+                }
+                None
+            }
+            // If we need to satisfy both of the formulas, just return the largest distance found between the two
+            RangedPhi::And(lhs, rhs) => {
+                if let Some(lhs_distance) = self.visit_ranged_phi(lhs, state) {
+                    if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
+                        return Some(Ord::max(lhs_distance, rhs_distance));
+                    }
+                }
+                None
+            }
+            // Iterate through all entires in the hashmap mapping symbols to ranges
+            // Find distance for each symbol in the current state, to the closest acceptance region,
+            // Add all distances and return
+            RangedPhi::Proposition(proposition_range) => {
+                let mut cumulative_distance = 0;
+
+                for (symbol, range) in proposition_range {
+                    if let Some(state_of_symbol) = state.0.get(symbol) {
+                        let res = self.distance_to_range_bound(range, state_of_symbol);
+                        cumulative_distance = cumulative_distance + res;
+                    }
+                }
+
+                Some(cumulative_distance)
+            }
+            RangedPhi::True => Some(0),
+            RangedPhi::False => None,
```

Comment:
This should be `Some(0)`. Otherwise `x < 5 || false` will also get a `None` distance.

Reply:
This case should be handled in `visit_ranged_phi` in the `Or` case.

If either of the lhs or rhs is None, the other is returned (given that it is some), else the min of the both is returned. 
```
RangedPhi::Or(lhs, rhs) => {
                if let Some(lhs_distance) = self.visit_ranged_phi(lhs, state) {
                    return if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
                        Some(Ord::min(lhs_distance, rhs_distance))
                    } else {
                        Some(lhs_distance)
                    };
                } else {
                    if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
                        return Some(rhs_distance);
                    }
                }
                None
            }
```

Reply:
Ah. Cool. Missed that. So None actually means undefined, and that makes sense.